### PR TITLE
[forms] Fix incorrect types + have form definition be reflected in what bound data is available

### PR DESCRIPTION
**Pull request by voxpelli**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/caolan/forms/blob/master/lib/forms.js
- [x] <del>If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.</del>

Discussed this briefly with maintainer of forms, @ljharb: https://twitter.com/ljharb/status/1346503976934510592

### CODE REVIEW

File: types/forms/index.d.ts [hunk 1]
```diff
@@ -1,7 +1,13 @@
 // Type definitions for forms 1.3
 // Project: https://github.com/caolan/forms
 // Definitions by: suXin <https://github.com/suXinjke>
+//                 Pelle Wessman <https://github.com/voxpelli>
```
Comment (ljharb) [inline types/forms/index.d.ts:4]:
```suggestion
//                 Pelle Wessman <https://github.com/voxpelli>
//                 Jordan Harband <https://github.com/ljharb>
```

this way i'll get pinged on PRs automatically :-)

File: types/forms/index.d.ts [hunk 2]
```diff
@@ -52,15 +58,15 @@ export interface FieldParameters {
 
 export type FieldIterator = (name: string, field: FieldBound) => string;
 
-export interface Field extends FieldParameters {
+export type Field<Data = any> = FieldParameters & {
```
Comment (ljharb) [inline types/forms/index.d.ts:61]:
```suggestion
export type Field<Data = unknown> = FieldParameters & {
```

File: types/forms/index.d.ts [hunk 3]
```diff
@@ -52,15 +58,15 @@ export interface FieldParameters {
...
     /** A widget object to use when rendering the field. */
     widget: Widget;
 
     /** Coerces the raw data from the request into the correct format for the field, returning the result, e.g. '123' becomes 123 for the number field. */
-    parse: (rawData: any) => any;
+    parse: (rawData: any) => Data;
```
Comment (ljharb) [inline types/forms/index.d.ts:66]:
```suggestion
    parse: (rawData: unknown) => Data;
```

File: types/forms/index.d.ts [hunk 4]
```diff
@@ -52,15 +58,15 @@ export interface FieldParameters {
...
 
     /** Returns a new bound field object. Calls parse on the data and stores in the bound field's data attribute, stores the raw value in the value attribute. */
-    bind: (rawData: any) => FieldBound;
+    bind: <RawData = any>(rawData: RawData) => FieldBound<Data, RawData>;
```
Comment (ljharb) [inline types/forms/index.d.ts:69]:
```suggestion
    bind: <RawData = unknown>(rawData: RawData) => FieldBound<Data, RawData>;
```

File: types/forms/index.d.ts [hunk 5]
```diff
@@ -79,14 +85,14 @@ export interface Field extends FieldParameters {
      * which returns a HTML representation of the field label, error message and widget wrapped in a div.
      */
     toHTML: (name?: string, iterator?: FieldIterator) => string;
-}
+};
 
-export interface FieldBound extends Field {
+export type FieldBound<Data = any, RawData = any> = Field<Data> & {
```
Comment (ljharb) [inline types/forms/index.d.ts:90]:
```suggestion
export type FieldBound<Data = unknown, RawData = unknown> = Field<Data> & {
```

File: types/forms/index.d.ts [hunk 6]
```diff
@@ -127,29 +133,33 @@ export interface FormFields {
     [key: string]: Field | FormFields;
 }
 
-export type FormHandleCallback = (form: Form) => void;
+export type FormHandleCallback<
+    Fields extends FormFields = FormFields,
+    Data extends (IncomingMessage | (Partial<FormData<Fields>> & { [key: string]: any })) = FormData<Fields>
```
Comment (ljharb) [inline types/forms/index.d.ts:138]:
```suggestion
    Data extends (IncomingMessage | (Partial<FormData<Fields>> & { [key: string]: unknown })) = FormData<Fields>
```
altho is the value of the indexed object really of type unknown, or is it a set of known types, or an array/object of those known types?

File: types/forms/index.d.ts [hunk 7]
```diff
@@ -127,29 +133,33 @@ export interface FormFields {
...
+    fields: Fields;
 
     /** Inspects a request or object literal and binds any data to the correct fields. */
-    handle: (
-        req: {
-            method: string,
-            url: string
-            body: string
-        },
+    handle: <Data extends IncomingMessage | (Partial<FormData<Fields>> & { [key: string]: any })>(
```
Comment (ljharb) [inline types/forms/index.d.ts:152]:
```suggestion
    handle: <Data extends IncomingMessage | (Partial<FormData<Fields>> & { [key: string]: unknown })>(
```
same q here

File: types/forms/index.d.ts [hunk 8]
```diff
@@ -127,29 +133,33 @@ export interface FormFields {
...
-            other?: FormHandleCallback
+            success?: FormHandleCallback<Fields, Data>
+            error?: FormHandleCallback<Fields, Data>
+            empty?: FormHandleCallback<Fields, Data>
         }
     ) => void;
 
     /** Binds data to correct fields, returning a new bound form object. */
-    bind: (data: any) => FormBound;
+    bind: <Data extends Partial<FormData<Fields>>>(data: (Data & { [key: string]: any }) | null | undefined) => FormBound<Fields, Data>;
```
Comment (ljharb) [inline types/forms/index.d.ts:162]:
```suggestion
    bind: <Data extends Partial<FormData<Fields>>>(data: (Data & { [key: string]: unknown }) | null | undefined) => FormBound<Fields, Data>;
```

File: types/forms/index.d.ts [hunk 9]
```diff
@@ -159,33 +169,33 @@ export interface Form {
     toHTML: (iterator?: FieldIterator) => string;
 }
 
-export interface FormBound extends Form {
+export interface FormBound<Fields extends FormFields = any, Data extends Partial<FormData<Fields>> = FormData<Fields>> {
```
Comment (ljharb) [inline types/forms/index.d.ts:172]:
```suggestion
export interface FormBound<Fields extends FormFields = unknown, Data extends Partial<FormData<Fields>> = FormData<Fields>> {
```

File: types/forms/index.d.ts [hunk 10]
```diff
@@ -159,33 +169,33 @@ export interface Form {
...
-    function array(params?: FieldParameters): Field;
-    function boolean(params?: FieldParameters): Field;
-    function date(params?: FieldParameters): Field;
-    function email(params?: FieldParameters): Field;
-    function number(params?: FieldParameters): Field;
-    function password(params?: FieldParameters): Field;
-    function string(params?: FieldParameters): Field;
-    function tel(params?: FieldParameters): Field;
-    function url(params?: FieldParameters): Field;
+    function array(params?: FieldParameters): Field<any[]>;
```
Comment (ljharb) [inline types/forms/index.d.ts:190]:
```suggestion
    function array<T = unknown>(params?: FieldParameters): Field<T[]>;
```

**Review by ljharb — COMMENTED** [review]
overall LGTM

(sorry for all the `unknown` comments, i have a thing against `any`)

File: types/forms/index.d.ts (hunk 9 above)
Comment (ljharb) [reply types/forms/index.d.ts:173]: ?

**Comment by voxpelli** [comment]
@ljharb I had to do two small tweaks to make the linting and tests happy: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/50461/commits/76531ae1a25962068f90b7115809abaaf2d5c1ba

**Review by ljharb — APPROVED** [review]

**Comment by voxpelli** [comment]
Ready to merge

**Pull request merged**